opm/simulators/wells/WellInterface_impl.hpp

```diff
@@ -1317,6 +1318,84 @@ namespace Opm
+        }
+        // for gas wells we may want to add a Forchheimer term if the WDFAC or WDFACCOR keyword is used
+        const auto& connection = this->well_ecl_.getConnections()[perf];
+        // viscosity is evaluated at bhp
+        const double mu = FluidSystem::gasPvt().viscosity(this->pvtRegionIdx(), ws.temperature, ws.bhp, getValue(intQuants.fluidState().Rv()), getValue(intQuants.fluidState().Rvw()));
+        const double phi = getValue(intQuants.porosity());
+        //double k = connection.Kh()/h * trans_mult;
+        double Kh = connection.Kh()* trans_mult;
+        double Ke = connection.Ke()* trans_mult;
+        double h = Kh / Ke;
+        double rw = connection.rw();
+        double rho = FluidSystem::referenceDensity(FluidSystem::gasPhaseIdx, this->pvtRegionIdx());
+        double scaling = 3.141592653589 * Kh;
+        double d = wdfac.useConnectionDFactor()? connection.dFactor() : wdfac.getDFactor(rho, mu, Ke, phi, rw, h);
+        const PhaseUsage& pu = this->phaseUsage();
+        double Q = std::abs(ws.perf_data.phase_rates[perf*pu.num_phases + pu.phase_pos[Gas]]);
+        return 1.0/(1.0/(trans_mult * connection.CF()) + (Q/2 * d / scaling));
+    }
+
+    template <typename TypeTag>
+    void
+    WellInterface<TypeTag>::
+    updateConnectionDFactor(const Simulator& simulator, SingleWellState& ws) const {
+
+        const auto& wdfac = this->well_ecl_.getWDFAC();
+        if (!wdfac.useDFactor()) {
+            return;
+        }
+        double rho = FluidSystem::referenceDensity(FluidSystem::gasPhaseIdx, this->pvtRegionIdx());
+        auto& perf_data = ws.perf_data;
+
+        for (int perf = 0; perf < this->number_of_perforations_; ++perf) {
+            const int cell_idx = this->well_cells_[perf];
+            const auto& intQuants = simulator.model().intensiveQuantities(cell_idx, /*timeIdx=*/ 0);
+            const double trans_mult = simulator.problem().template rockCompTransMultiplier<double>(intQuants, cell_idx);
+            // viscosity is evaluated at bhp
+            const double mu = FluidSystem::gasPvt().viscosity(this->pvtRegionIdx(), ws.temperature, ws.bhp, getValue(intQuants.fluidState().Rv()), getValue(intQuants.fluidState().Rvw()));
+            const double phi = getValue(intQuants.porosity());
+            const auto& connection = this->well_ecl_.getConnections()[perf];
+            //double h = connection.connectionLength();
+            //double k = connection.Kh()/h * trans_mult;
+            double Kh = connection.Kh()* trans_mult;
+            double Ke = connection.Ke()* trans_mult;
+            double h = Kh / Ke;
+            double rw = connection.rw();
+            double d = wdfac.useConnectionDFactor()? connection.dFactor() : wdfac.getDFactor(rho, mu, Ke, phi, rw, h);
+            double scaling = 3.141592653589 * Kh;
+            perf_data.connection_d_factor[perf] = d;
+            perf_data.connection_transmissibility_factor[perf] = this->well_index_[perf] * trans_mult;
```

Comment:
It's not clear to me if this breaks the `CTFAC` output of runs which have both `WDFAC` and `WELPI` for the same well.  I haven't managed to perform all of the requisite analysis, but I think it's possible you'll only get one of the updates (`WDFAC` or `WELPI`) but not both, and in that case it is not clear right now to me which update will "win".  I'd appreciate some comments here.

Reply:
Ops. Line 1381 should be removed here. I moved it to a separate function and forgot to remove it from here.